File: myutils/src/main/java/com/jty/myutils/utils/DensityUtils.java
package com.jty.myutils.utils;

import android.content.Context;
import android.util.TypedValue;

/** 
 * 常用单位转换辅助类
 * @author jty 
 */

public class DensityUtils {
	
	private DensityUtils(){
		//cannot be instantiated
		throw new UnsupportedOperationException("cannot be instantiated");
	}
	
	/** 
     * dp转px 
     *  
     * @param context 
     * @param dp
     * @return px
     */  
	public static int dp2px(Context context,float dp){
		return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,
                dp, context.getResources().getDisplayMetrics());
		
	}
	/** 
     * sp转px 
     *  
     * @param context 
     * @param sp
     * @return px
     */ 
	public static int sp2px(Context context,float sp){
		return (int)TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP,
                sp, context.getResources().getDisplayMetrics());
	}
	/** 
     * px转dp 
     *  
     * @param context 
     * @param px
     * @return dp
     */ 
	public static float px2dp(Context context,float px){
		final float scale=context.getResources().getDisplayMetrics().density;
		return (px/scale);
	}
	/** 
     * px转sp 
     *  
     * @param context
     * @param px
	 * @return sp
     */
//	@Deprecated
	public static float px2sp(Context context,float px){
		return (px/context.getResources().getDisplayMetrics().scaledDensity);
	}
	
	
	
	
}

File: mykotlinutils/src/main/java/com/jty/mykotlinutils/utils/DensityUtils.kt
package com.jty.mykotlinutils.utils

import android.content.Context
import android.util.TypedValue

/**
 * Created by TaoYuan on 2017/9/20 0020.
 *
 * Screen Density Convert
 *
 * This Class Cannot be Instantiated,
 * Direct Use Static Method Call It In Java.
 * Just Like This : DensityUtils.Companion.dp2px(context,dp);
 */

class DensityUtils private constructor() {

    init {
        //cannot be instantiated
        throw UnsupportedOperationException("cannot be instantiated")
    }

    companion object {

        /**
         * dp转px
         *
         * @param context
         * @param dp
         * @return px
         */
        fun dp2px(context: Context, dp: Float): Int = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,
                dp, context.resources.displayMetrics).toInt()

        /**
         * sp转px
         *
         * @param context
         * @param sp
         * @return px
         */
        fun sp2px(context: Context, sp: Float): Int = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP,
                sp, context.resources.displayMetrics).toInt()

        /**
         * px转dp
         *
         * @param context
         * @param px
         * @return dp
         */
        fun px2dp(context: Context, px: Float): Float =
                px / context.resources.displayMetrics.density

        /**
         * px转sp
         *
         * @param context
         * @param px
         * @return sp
         */
        fun px2sp(context: Context, px: Float): Float =
                px / context.resources.displayMetrics.scaledDensity
    }


}
